File: heron/lc308.java
public class NumMatrix {

    int[][] mMatrix;

    int m;
    int n;

    public NumMatrix(int[][] matrix) {

        // Returns if the input matrix is empty
        if (matrix == null || matrix.length == 0 || matrix[0].length == 0) {
            return;
        }

        // Stores the class properties
        mMatrix = matrix;
        m = matrix.length;
        n = matrix[0].length;

        // Calculates matrix[i][j] = matrix[i][0] + matrix[i][1] ... + matrix[i][j]
        for (int i=0; i<m; i++) {
            for (int j=1; j<n; j++) {
                mMatrix[i][j] += mMatrix[i][j-1];
            }
        }

    }

    // O(n)
    public void update(int row, int col, int val) {

        int originalValue = (col == 0) ? mMatrix[row][0] : mMatrix[row][col] - mMatrix[row][col-1];
        int delta = val - originalValue;

        for (int j=col; j<n; j++) {
            mMatrix[row][j] += delta;
        }

    }

    // O(m)
    public int sumRegion(int row1, int col1, int row2, int col2) {
        int result = 0;
        for (int i=row1; i<=row2; i++) {
            result += (col1 == 0) ? mMatrix[i][col2] : mMatrix[i][col2] - mMatrix[i][col1-1];
        }
        return result;
    }

}

File: alice/LC308.py
class NumMatrix(object):

    def __init__(self, matrix):
        """
        :type matrix: List[List[int]]
        """

        if not matrix:
            return

        self.row = len(matrix)
        self.col = len(matrix[0])
        self.matrix = [[0] * self.col for _ in range(self.row)]
        for i in xrange(self.row):
            for j in xrange(self.col):
                self.matrix[i][j] = self.matrix[max(0, i - 1)][j] + matrix[i][j]

    def update(self, row, col, val):
        """
        :type row: int
        :type col: int
        :type val: int
        :rtype: void
        """

        origin = self.matrix[0][col] if not row else self.matrix[row][col] - self.matrix[row - 1][col]
        for i in range(row, self.row):
            self.matrix[i][col] += (val - origin)

    def sumRegion(self, row1, col1, row2, col2):
        """
        :type row1: int
        :type col1: int
        :type row2: int
        :type col2: int
        :rtype: int
        """

        bottom = sum([self.matrix[row2][j] for j in range(col1, col2 + 1)])
        top = 0 if not row1 else sum([self.matrix[row1 - 1][j] for j in range(col1, col2 + 1)])

        return bottom - top


# Your NumMatrix object will be instantiated and called as such:
# obj = NumMatrix(matrix)
# obj.update(row,col,val)
# param_2 = obj.sumRegion(row1,col1,row2,col2)